Fix PHPStan for Twig 4.x

VincentLanglet opened the pull request:
Hi @fabpot, to help introducing PHPStan on 4.x, I made the merge 3.x into 4.x
- Fixing the conflits
- Fixing PHPStan build

I have
- A commit with the merge https://github.com/twigphp/Twig/pull/4466/commits/6a82e16617183c793bad4cfa672cee3c3923b714

- A commit with the update needed to get PHPStan build green in 4.x https://github.com/twigphp/Twig/pull/4466/commits/cb89b7528c997dd11a1a28192eba11a9fdbe5f5f

VincentLanglet committed:
PHPStan fixes

VincentLanglet commented on src/Environment.php:
```diff
@@ -54,7 +55,7 @@ class Environment
     private LoaderInterface $loader;
     private bool $debug;
     private bool $autoReload;
-    private CacheInterface|string|false $cache;
```
Type was wrong, it's always CacheInterface unlike the originalCache.

VincentLanglet commented on src/Node/EmptyNode.php:
```diff
@@ -26,7 +26,7 @@ public function __construct(int $lineno = 0)
     }
 
-    public function setNode(string $name, Node $node): void
+    public function setNode(string|int $name, Node $node): void
```
Parent accepts string|int now.

VincentLanglet commented on src/Runtime/LoopIterator.php:
```diff
@@ -45,11 +45,14 @@ public function __construct($seq)
                 $seq = $seq->getIterator();
             } while ($seq instanceof \IteratorAggregate);
             $this->seq = $seq;
-        } elseif (is_iterable($seq)) {
```
Is iterable is not always an Iterator but a Traversable.

So we technically have to use IteratorIterator if it's not an Iterator.

fabpot replied:
Can you add a test that would break without this change?

fabpot commented:
@VincentLanglet I've just merged 3.x into 4.x to keep this PR only about phpstan.

fabpot commented on src/ExpressionParser.php:
```diff
@@ -51,9 +49,9 @@ class ExpressionParser
+    /** @var array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractExpression>}> */
     private array $unaryOperators;
-    /** @var array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractBinary>, associativity: self::OPERATOR_*}> */
+    /** @var array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractExpression>, associativity: self::OPERATOR_*}> */
```
These 2 changes look wrong to me.

VincentLanglet replied:
Those values are set in construct
```
 $this->unaryOperators = $env->getUnaryOperators();
$this->binaryOperators = $env->getBinaryOperators();
```

They call the method on `$this->extensionSet`.
The ExtensionSet gets them from multiple ExtensionInterface, 
and getOperators is
```
/**
     * Returns a list of operators to add to the existing list.
     *
     * @return array<array> First array of unary operators, second array of binary operators
     *
     * @psalm-return array{
     *     array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class: class-string<AbstractExpression>}>,
     *     array<string, array{precedence: int, precedence_change?: OperatorPrecedenceChange, class?: class-string<AbstractExpression>, associativity: ExpressionParser::OPERATOR_*}>
     * }
     */
    public function getOperators(): array;
```

This is because 
```
'??' => ['precedence' => 5, 'class' => NullCoalesceExpression::class, 'associativity' => ExpressionParser::OPERATOR_RIGHT],
```
is not an AbstractBinary but just an AbtractExpression.

fabpot replied:
Indeed. I suppose all binary operators should extend AbstractBinary. I'm going to have a look at it.

VincentLanglet replied:
> Indeed. I suppose all binary operators should extend AbstractBinary. I'm going to have a look at it.

The issue already exists on 3.x if you want to take a look and fix it on the lower branch.
You can ping me after the fix, I'll update the PHPDoc if needed :)

fabpot replied:
See #4482

fabpot replied:
This PR can now be rebased.

VincentLanglet commented:
> Can you add a test that would break without this change?

For the iterable/Iterator issue, I think it's an issue with PHPStan instead.
So I created an issue on https://github.com/phpstan/phpstan/issues/12102 (and a fix on phpstan side)

fabpot commented:
> > Can you add a test that would break without this change?
> 
> For the iterable/Iterator issue, I think it's an issue with PHPStan instead. So I created an issue on [phpstan/phpstan#12102](https://github.com/phpstan/phpstan/issues/12102) (and a fix on phpstan side)

Let's revert the change then.

VincentLanglet commented:
> > > Can you add a test that would break without this change?
> > 
> > 
> > For the iterable/Iterator issue, I think it's an issue with PHPStan instead. So I created an issue on [phpstan/phpstan#12102](https://github.com/phpstan/phpstan/issues/12102) (and a fix on phpstan side)
> 
> Let's revert the change then.

I reverted the extra if I added.

We still need `/** @var \Iterator<TKey, TValue> $seq */` to not lose the `template type`.

VincentLanglet commented:
I rebased @fabpot

fabpot commented:
Thank you @VincentLanglet.
